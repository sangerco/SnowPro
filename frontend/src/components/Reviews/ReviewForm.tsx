import React, { useEffect, useState } from "react";
import { connect } from 'react-redux';
import { useAuth } from "../AuthProvider";
import { sendNewReviewData } from "../../oldRedux/actions/reviewActions";
import { fetchTagData } from "../../oldRedux/actions/tagActions";
import NewTagForm from "../Tags/NewTagForm";
import { NewTagData } from "../../oldRedux/types/tagTypes";
import { RootState } from "../../oldRedux/store";
import { Button, Container, Divider, Dropdown, Form, Message, TextArea } from 'semantic-ui-react';
import { TagData } from "../../oldRedux/types/tagTypes";
import { useParams } from "react-router";

interface NewReviewData {
    userId: string;
    skiAreaSlug: string;
    header: string;
    body: string;
    stars: number;
    photos: string[];
    tagIds: string[];
};

interface NewReviewProps {
    newReview: NewReviewData | null;
    error: string;
    tags: TagData[] | null;
    sendNewReviewData: ( userId: string,
                            skiAreaSlug: string,
                            header: string,
                            body: string,
                            stars: number,
                            photos: string[],
                            tagIds: string[]) => Promise<void>;
    fetchTagData: () => Promise<void>
};

const ReviewForm: React.FC<NewReviewProps> = ({ newReview, error, tags, sendNewReviewData, fetchTagData }) => {
    const { skiAreaSlug } = useParams<{ skiAreaSlug: string }>();
    const { userId } = useAuth();
    const initialReviewState = { userId: userId ?? '',
                                    skiAreaSlug: skiAreaSlug,
                                    header: '',
                                    body: '',
                                    stars: 0,
                                    photos: [],
                                    tagIds: []}

    const [formData, setFormData] = useState(initialReviewState);
    // const [tags, setTags] = useState<TagData[]>([]);
    const [photoLinks, setPhotoLinks] = useState<string[]>(['']);
    const [showCreateNewTagForm, setShowCreateNewTagForm] = useState<boolean>(false);
    
    useEffect(() => {
        const fetchData = async () => {
            try {
                await fetchTagData();
            } catch (error: any) {
                return `Can't retrieve tag data: error ${error.message}`
            }
        };
        fetchData();
    }, [fetchTagData]);

    const dropdownOptions = [
        { key: 1, text: '1', value: 1 },
        { key: 2, text: '2', value: 2 },
        { key: 3, text: '3', value: 3 },
        { key: 4, text: '4', value: 4 },
        { key: 5, text: '5', value: 5 },
    ]

    const tagOptions = (tags || []).map((tag) => ({ key: tag.id, text: tag.tag, value: tag.id}))

    console.log(tagOptions);

    const handleSubmit = async (e: React.FormEvent<HTMLFormElement>) => {
        e.preventDefault();

        await sendNewReviewData(formData.userId,
                                    formData.skiAreaSlug ?? '',
                                    formData.header,
                                    formData.body,
                                    formData.stars,
                                    formData.photos,
                                    formData.tagIds);
                
        setFormData(initialReviewState);
    };

    const handleChange: React.ChangeEventHandler<HTMLInputElement> = (e) => {
        const { name, value } = e.target;
        setFormData({
            ...formData,
            [name]: value
        });
    };

    const handlePhotoChange = (index: number, value: string) => {
        const updatedPhotoLinks = [...photoLinks];
        updatedPhotoLinks[index] = value;
        setPhotoLinks(updatedPhotoLinks);
    };

    const handleAddPhotoInput = () => {
        setPhotoLinks([...photoLinks, '']);
    };

    const handleRemovePhotoInput = (index: number) => {
        const updatedPhotoLinks = [...photoLinks];
        updatedPhotoLinks.splice(index, 1);
        setPhotoLinks(updatedPhotoLinks);
    };

    const initialNewTagData: NewTagData = {
        tag: 'New Tag'
    };

    return (
        <Container fluid>
            <Form onSubmit={handleSubmit} error>
                <Form.Field>
                    <label>Header</label>
                    <input name="header" value={formData.header} onChange={handleChange} />
                </Form.Field>
                <Form.Field
                    id='form-textarea-control-review'
                    control={TextArea}
                    label='Review'
                    placeholder='How Was It?'
                    name='body'
                    value={formData.body}
                    onChange={handleChange}
                />
                {photoLinks.map((photoLink, index) => (
                    <Form.Field key={index}>
                        <label>Photo Links</label>
                        <input 
                            type="text" 
                            value={photoLink} 
                            onChange={(e) => handlePhotoChange(index, e.target.value)} 
                        />
                        {index === photoLinks.length - 1 && (
                            <Button icon='plus' onClick={handleAddPhotoInput} />
                        )}
                        {(index > 0 && (
                            <Button icon='minus' onClick={() => handleRemovePhotoInput(index)} />
                        ))}
                    </Form.Field>
                ))}
                <Button type="submit">Send It!</Button>
                <Message 
                    error
                    header={error}
                    content={error}
                />
            </Form>
            <Divider />
            <label>Tags</label>
            <Dropdown clearable options={tagOptions} multiple fluid selection value={formData.tagIds} />
            <Button onClick={() => setShowCreateNewTagForm(true)} size="small">Create New Tag?</Button>
            {showCreateNewTagForm && <NewTagForm newTag={initialNewTagData} />}
            <Divider />
            <label>How would you rate your experience?</label>
            <Dropdown clearable options={dropdownOptions} selection fluid value={formData.stars} />
            <Divider />
        </Container>
    )
};

const mapStateToProps = (state: RootState) => ({
    tags: state.tag.data,
    tagError: state.tag.error,
    newReview: state.newReview.data,
    reviewError: state.newReview.error
});

const mapDispatchToProps = {
    sendNewReviewData,
    fetchTagData
}

export default connect(mapStateToProps, mapDispatchToProps)(ReviewForm)