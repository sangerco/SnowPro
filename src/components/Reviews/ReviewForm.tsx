import React, { useEffect, useState } from "react";
import { createReview, NewReviewData } from "../../redux/slices/reviewSlice";
import { Button, Form } from "semantic-ui-react";
import { RootState, AppDispatch } from "../../redux/store";
import { useDispatch, useSelector } from "react-redux";
import { useParams } from "react-router";

const ReviewForm: React.FC = () => {
  const { slug } = useParams();
  const dispatch = useDispatch<AppDispatch>();
  const auth = useSelector((state: RootState) => state.auth);
  const userId = auth.data?.id;
  const username = auth.data?.username;

  const initialReviewState: NewReviewData = {
    userId: userId ?? "",
    username: username ?? "",
    skiAreaSlug: slug ?? "",
    header: "",
    body: "",
    stars: 0,
    photos: [],
    tags: [],
  };

  const [formData, setFormData] = useState(initialReviewState);
  const [photos, setPhotos] = useState(initialReviewState.photos);
  const [tags, setTags] = useState(initialReviewState.tags);

  const handleSubmit = (e: React.FormEvent<HTMLFormElement>) => {
    e.preventDefault();

    dispatch(createReview(formData));

    setFormData(initialReviewState);
  };

  const handleInputChange: React.ChangeEventHandler<HTMLInputElement> = (e) => {
    const { name, value } = e.target;

    setFormData({
      ...formData,
      [name]: value,
    });
  };

  const handleTextAreaChange: React.ChangeEventHandler<HTMLTextAreaElement> = (
    e
  ) => {
    const { name, value } = e.target;

    setFormData({
      ...formData,
      [name]: value,
    });
  };

  const handlePhotoChange = (index: number, value: string) => {
    if (formData.photos) {
      const updatedPhotoLinks = [...formData.photos];
      updatedPhotoLinks[index] = value;
      setPhotos(updatedPhotoLinks);
    }
  };

  const handleAddPhotoInput = () => {
    if (formData.photos) {
      setPhotos([...formData.photos, ""]);
    }
  };

  const handleRemovePhotoInput = (index: number) => {
    if (formData.photos) {
      const updatedPhotoLinks = [...formData.photos];
      updatedPhotoLinks.splice(index, 1);
      setPhotos(updatedPhotoLinks);
    }
  };
};
