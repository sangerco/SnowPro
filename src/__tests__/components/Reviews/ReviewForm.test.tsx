/* eslint-disable testing-library/no-node-access */
import React from "react";
import ReviewForm from "../../../components/Reviews/ReviewForm";
import { render, fireEvent, screen, waitFor } from "@testing-library/react";
import { Provider } from "react-redux";
import { BrowserRouter } from "react-router-dom";
import configureStore from "redux-mock-store";
import thunk from "redux-thunk";
import nock, { pendingMocks } from "nock";

const middlewares = [thunk];
const mockStore = configureStore(middlewares);

describe("test Review Form", () => {
  const initialState = {
    auth: {
      data: {
        id: "11",
      },
    },
    reviews: {
      reviews: [
        {
          id: "review-1",
          userId: "11",
          username: "testuser",
          skiAreaSlug: "ski-area-1",
          skiAreaName: "Ski Area 1",
          header: "test review",
          body: "test review body",
          stars: 5,
          photos: [],
          createdAt: "fake created Date",
          replies: [],
        },
        {
          id: "review-2",
          userId: "22",
          username: "testuser2",
          skiAreaSlug: "ski-area-2",
          skiAreaName: "Ski Area 2",
          header: "test review",
          body: "test review body",
          stars: 2,
          photos: [],
          createdAt: "fake created Date",
          replies: [],
        },
      ],
    },
    skiAreas: {
      skiArea: {
        slug: "ski-area-1",
        name: "Ski Area 1",
        country: "US",
        region: "CO",
        href: "www.fakehref.com",
        location: {
          latitude: 0,
          longitude: 0,
        },
        lifts: {},
        twitter: {},
      },
      skiAreas: [
        {
          slug: "ski-area-1",
          name: "Ski Area 1",
          country: "US",
          region: "CO",
          href: "www.fakehref.com",
          location: {
            latitude: 0,
            longitude: 0,
          },
          lifts: {},
          twitter: {},
        },
        {
          slug: "ski-area-1",
          name: "Ski Area 1",
          country: "US",
          region: "CO",
          href: "www.fakehref.com",
          location: {
            latitude: 0,
            longitude: 0,
          },
          lifts: {},
          twitter: {},
        },
      ],
    },
  };
  const store = mockStore(initialState);
  it("should render without crashing", () => {
    render(
      <Provider store={store}>
        <BrowserRouter>
          <ReviewForm />
        </BrowserRouter>
      </Provider>
    );
  });
  it("should match snapshot", () => {
    const { asFragment } = render(
      <Provider store={store}>
        <BrowserRouter>
          <ReviewForm />
        </BrowserRouter>
      </Provider>
    );
    expect(asFragment()).toMatchSnapshot();
  });
});

jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useParams: jest.fn(),
}));

describe("test create review", () => {
  const skiAreaSlug = "ski-area-1";
  const useParamsMock = jest.requireMock("react-router-dom")
    .useParams as jest.Mock;
  useParamsMock.mockReturnValue({ skiAreaSlug });
  const initialState = {
    auth: {
      data: {
        id: "11",
      },
    },
    reviews: {
      reviews: [
        {
          id: "review-1",
          userId: "11",
          username: "testuser",
          skiAreaSlug: "ski-area-1",
          skiAreaName: "Ski Area 1",
          header: "test review",
          body: "test review body",
          stars: 5,
          photos: [],
          createdAt: "fake created Date",
          replies: [],
        },
        {
          id: "review-2",
          userId: "22",
          username: "testuser2",
          skiAreaSlug: "ski-area-2",
          skiAreaName: "Ski Area 2",
          header: "test review",
          body: "test review body",
          stars: 2,
          photos: [],
          createdAt: "fake created Date",
          replies: [],
        },
      ],
    },
    skiAreas: {
      skiArea: {
        slug: "ski-area-1",
        name: "Ski Area 1",
        country: "US",
        region: "CO",
        href: "www.fakehref.com",
        location: {
          latitude: 0,
          longitude: 0,
        },
        lifts: {},
        twitter: {},
      },
      skiAreas: [
        {
          slug: "ski-area-1",
          name: "Ski Area 1",
          country: "US",
          region: "CO",
          href: "www.fakehref.com",
          location: {
            latitude: 0,
            longitude: 0,
          },
          lifts: {},
          twitter: {},
        },
        {
          slug: "ski-area-1",
          name: "Ski Area 1",
          country: "US",
          region: "CO",
          href: "www.fakehref.com",
          location: {
            latitude: 0,
            longitude: 0,
          },
          lifts: {},
          twitter: {},
        },
      ],
    },
  };
  const store = mockStore(initialState);
  it("should render the form", async () => {
    nock("http://localhost:5000")
      .defaultReplyHeaders({
        "access-control-allow-origin": "*",
      })
      .post(`/api/ski-areas/ski-area-1/review`, {
        skiAreaSlug: "ski-area-1",
      })
      .reply(201, {
        review: {
          id: "review-3",
          userId: "33",
          skiAreaSlug: "ski-area-1",
          header: "test review",
          body: "test review body",
          stars: 3,
          createdAt: "fake created date",
        },
      });
    render(
      <Provider store={store}>
        <BrowserRouter>
          <ReviewForm />
        </BrowserRouter>
      </Provider>
    );

    await waitFor(() => {
      const reviewForm = screen.getByTestId("review-form");
      expect(reviewForm).toBeInTheDocument();
    });

    await waitFor(() => {
      const reviewBody = screen.getByPlaceholderText("How was it?");
      expect(reviewBody).toBeInTheDocument();
    });

    await waitFor(() => {
      const submitButton = screen.getByTestId("review-button");
      expect(submitButton).toBeInTheDocument();
    });
  });

  it("should submit the form", async () => {
    nock("http://localhost:5000")
      .defaultReplyHeaders({
        "access-control-allow-origin": "*",
      })
      .post(`/api/ski-areas/ski-area-1/review`, {
        skiAreaSlug: "ski-area-1",
      })
      .reply(201, {
        review: {
          id: "review-3",
          userId: "33",
          skiAreaSlug: "ski-area-1",
          header: "test review",
          body: "test review body",
          stars: 3,
          createdAt: "fake created date",
        },
      });
    render(
      <Provider store={store}>
        <BrowserRouter>
          <ReviewForm />
        </BrowserRouter>
      </Provider>
    );

    expect(pendingMocks().length).toBe(2);

    const body = screen.getByPlaceholderText("How was it?");
    fireEvent.change(body, { target: { value: "test review" } });

    const reviewForm = await screen.findByTestId("review-form");
    fireEvent.submit(reviewForm);
  });
});
