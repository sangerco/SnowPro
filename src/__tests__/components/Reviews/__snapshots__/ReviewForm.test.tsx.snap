// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test Review Form should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="ui grid"
  >
    <div
      class="row"
    >
      <div
        class="four wide column"
      >
        <div
          class="ui segment"
          style="overflow: auto; max-height: 75vh;"
        >
          <div
            class="ui card"
            id="ski-area-card"
          >
             
            <div
              class="content"
            >
              <a
                href="/ski-areas/ski-area-1"
              >
                Ski Area 1
              </a>
            </div>
            <div
              class="content"
            >
              <a
                href="/ski-areas/ski-area-1"
              >
                Ski Area 1
              </a>
            </div>
          </div>
        </div>
      </div>
      <div
        class="eight wide column"
      >
        <div
          class="ui segment"
        >
          <div
            class="ui fluid container"
          >
            <form
              class="ui form"
              data-testid="review-form"
            >
              <div
                class="required field"
              >
                <label>
                  Header
                </label>
                <input
                  name="header"
                  value=""
                />
              </div>
              <div
                class="required field"
              >
                <textarea
                  name="body"
                  placeholder="How was it?"
                  required=""
                  rows="3"
                />
              </div>
               
              <label>
                How would you rate your experience?
              </label>
              <div
                aria-expanded="false"
                class="ui fluid selection dropdown"
                role="listbox"
                tabindex="0"
              >
                <i
                  aria-hidden="true"
                  class="dropdown icon clear"
                />
                <div
                  class="menu transition"
                >
                  <div
                    aria-checked="false"
                    aria-selected="true"
                    class="selected item"
                    role="option"
                    style="pointer-events: all;"
                  >
                    <span
                      class="text"
                    >
                      1
                    </span>
                  </div>
                  <div
                    aria-checked="false"
                    aria-selected="false"
                    class="item"
                    role="option"
                    style="pointer-events: all;"
                  >
                    <span
                      class="text"
                    >
                      2
                    </span>
                  </div>
                  <div
                    aria-checked="false"
                    aria-selected="false"
                    class="item"
                    role="option"
                    style="pointer-events: all;"
                  >
                    <span
                      class="text"
                    >
                      3
                    </span>
                  </div>
                  <div
                    aria-checked="false"
                    aria-selected="false"
                    class="item"
                    role="option"
                    style="pointer-events: all;"
                  >
                    <span
                      class="text"
                    >
                      4
                    </span>
                  </div>
                  <div
                    aria-checked="false"
                    aria-selected="false"
                    class="item"
                    role="option"
                    style="pointer-events: all;"
                  >
                    <span
                      class="text"
                    >
                      5
                    </span>
                  </div>
                </div>
              </div>
              <div
                class="ui divider"
              />
              <button
                class="ui button"
                data-testid="review-button"
                type="submit"
              >
                Send it!
              </button>
              <div
                class="ui divider"
              />
            </form>
          </div>
        </div>
      </div>
      <div
        class="four wide column"
      >
        <div
          class="ui card"
        >
          <div
            class="content"
            id="review-card"
          >
            <div
              class="header"
            >
              <a
                href="/ski-areas/reviews/review-1"
              >
                test review
              </a>
               
              <div
                class="ui divider"
              />
              Ski Area 1
            </div>
            <div
              class="description"
            >
              By testuser
            </div>
            <div
              class="meta"
            >
              <div
                class="ui star disabled rating"
                role="radiogroup"
                tabindex="0"
              >
                <i
                  aria-checked="false"
                  aria-posinset="1"
                  aria-setsize="5"
                  class="active icon"
                  role="radio"
                  tabindex="-1"
                />
                <i
                  aria-checked="false"
                  aria-posinset="2"
                  aria-setsize="5"
                  class="active icon"
                  role="radio"
                  tabindex="-1"
                />
                <i
                  aria-checked="false"
                  aria-posinset="3"
                  aria-setsize="5"
                  class="active icon"
                  role="radio"
                  tabindex="-1"
                />
                <i
                  aria-checked="false"
                  aria-posinset="4"
                  aria-setsize="5"
                  class="active icon"
                  role="radio"
                  tabindex="-1"
                />
                <i
                  aria-checked="true"
                  aria-posinset="5"
                  aria-setsize="5"
                  class="active icon"
                  role="radio"
                  tabindex="-1"
                />
              </div>
            </div>
            Invalid Date
          </div>
        </div>
        <div
          class="ui card"
        >
          <div
            class="content"
            id="review-card"
          >
            <div
              class="header"
            >
              <a
                href="/ski-areas/reviews/review-2"
              >
                test review
              </a>
               
              <div
                class="ui divider"
              />
              Ski Area 2
            </div>
            <div
              class="description"
            >
              By testuser2
            </div>
            <div
              class="meta"
            >
              <div
                class="ui star disabled rating"
                role="radiogroup"
                tabindex="0"
              >
                <i
                  aria-checked="false"
                  aria-posinset="1"
                  aria-setsize="5"
                  class="active icon"
                  role="radio"
                  tabindex="-1"
                />
                <i
                  aria-checked="true"
                  aria-posinset="2"
                  aria-setsize="5"
                  class="active icon"
                  role="radio"
                  tabindex="-1"
                />
                <i
                  aria-checked="false"
                  aria-posinset="3"
                  aria-setsize="5"
                  class="icon"
                  role="radio"
                  tabindex="-1"
                />
                <i
                  aria-checked="false"
                  aria-posinset="4"
                  aria-setsize="5"
                  class="icon"
                  role="radio"
                  tabindex="-1"
                />
                <i
                  aria-checked="false"
                  aria-posinset="5"
                  aria-setsize="5"
                  class="icon"
                  role="radio"
                  tabindex="-1"
                />
              </div>
            </div>
            Invalid Date
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
